dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1727,6 +1728,25 @@ private List<Event> inlineLLVMThreadLocal(FunctionCall call) {
         );
     }
 
+    private List<Event> inlineFfs(FunctionCall call) {
+        //see https://linux.die.net/man/3/ffs
+        final String name = call.getCalledFunction().getName();
+        checkArgument(call.getArguments().size() == 1,
+                "Expected 1 parameter for \"%s\", got %s.", name, call.getArguments().size());
+        final Expression input = call.getArguments().get(0);
+        final Register resultReg = getResultRegister(call);
+        final Type type = resultReg.getType();
+        checkArgument(type instanceof IntegerType,
+                "Non-integer %s type for \"%s\".", name, type);
+        final Expression cttz = expressions.makeCTTZ(input);
+        final IntegerType cttzType = (IntegerType)cttz.getType();
+        final Expression widthExpr = expressions.makeValue(BigInteger.valueOf(((IntegerType)type).getBitWidth()), cttzType);
+        final Expression count = expressions.makeAdd(cttz, expressions.makeOne(cttzType));
+        final Expression ite = expressions.makeITE(expressions.makeEQ(cttz, widthExpr), expressions.makeZero(cttzType), count);
+        final Event assignment = EventFactory.newLocal(resultReg, ite);
+        return List.of(assignment);
```

Comment:
The `widthExpr` should use `cttzType.getBitWidth()`, then `type` becomes obsolete.
I would also rename `cttzType` to `inputType` and define it as `input.getType()` before constructing `cttz`.
If you add a cast from `inputType/cttzType` to `type` (`outputType` might be a good name) at the very end, then this method could also handle `ffsl` and `ffsll`.

Reply:
The cast should be unsigned. Otherwise, LGTM.